File: swift-bridge/Sources/SceneKitBridge/HitTest.swift
import CoreGraphics
import Foundation
import SceneKit

final class SceneHitTestResultsBox: NSObject {
    let results: [SCNHitTestResult]

    init(results: [SCNHitTestResult]) {
        self.results = results
    }
}

func scnBorrowHitTestResultsBox(_ handle: UnsafeMutableRawPointer?) -> SceneHitTestResultsBox? {
    scnBorrow(handle)
}

@_cdecl("scn_view_hit_test")
public func scn_view_hit_test(
    _ viewHandle: UnsafeMutableRawPointer?,
    _ x: Double,
    _ y: Double
) -> UnsafeMutableRawPointer? {
    guard let view: SCNView = scnBorrow(viewHandle) else { return nil }
    let results = view.hitTest(CGPoint(x: x, y: y), options: nil)
    return scnRetain(SceneHitTestResultsBox(results: results))
}

@_cdecl("scn_hit_test_results_count")
public func scn_hit_test_results_count(_ resultsHandle: UnsafeMutableRawPointer?) -> Int {
    guard let results = scnBorrowHitTestResultsBox(resultsHandle) else { return 0 }
    return results.results.count
}

@_cdecl("scn_hit_test_results_get")
public func scn_hit_test_results_get(
    _ resultsHandle: UnsafeMutableRawPointer?,
    _ index: Int
) -> UnsafeMutableRawPointer? {
    guard let results = scnBorrowHitTestResultsBox(resultsHandle),
          index >= 0,
          index < results.results.count
    else { return nil }
    return scnRetain(results.results[index])
}

@_cdecl("scn_hit_test_result_node")
public func scn_hit_test_result_node(_ resultHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let result: SCNHitTestResult = scnBorrow(resultHandle) else { return nil }
    return scnRetain(result.node)
}

@_cdecl("scn_hit_test_result_world_coordinates")
public func scn_hit_test_result_world_coordinates(
    _ resultHandle: UnsafeMutableRawPointer?,
    _ outVector: UnsafeMutableRawPointer?
) -> Bool {
    guard let result: SCNHitTestResult = scnBorrow(resultHandle) else { return false }
    return scnWriteVector3(result.worldCoordinates, out: outVector)
}

File: src/hit_test.rs
use crate::ffi;
use crate::math::Vector3;
use crate::node::Node;
use crate::private::handle_type;
use crate::view::View;

handle_type!(HitTestResult);
handle_type!(HitTestResults, "NSArray<SCNHitTestResult>");

impl HitTestResults {
    /// Mirrors `NSArray<SCNHitTestResult>.count`.
    #[must_use]
    pub fn count(&self) -> usize {
        unsafe { ffi::scn_hit_test_results_count(self.ptr) }
    }

    /// Returns the `NSArray<SCNHitTestResult>.isEmpty` value.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.count() == 0
    }

    /// Mirrors `NSArray<SCNHitTestResult>.get`.
    #[must_use]
    pub fn get(&self, index: usize) -> Option<HitTestResult> {
        unsafe { HitTestResult::from_raw(ffi::scn_hit_test_results_get(self.ptr, index)) }
    }

    /// Mirrors `NSArray<SCNHitTestResult>.first`.
    #[must_use]
    pub fn first(&self) -> Option<HitTestResult> {
        self.get(0)
    }
}

impl HitTestResult {
    /// Mirrors `SCNHitTestResult.node`.
    #[must_use]
    pub fn node(&self) -> Option<Node> {
        unsafe { Node::from_raw(ffi::scn_hit_test_result_node(self.ptr)) }
    }

    /// Mirrors `SCNHitTestResult.worldCoordinates`.
    #[must_use]
    pub fn world_coordinates(&self) -> Option<Vector3> {
        let mut value = Vector3::default();
        let ok = unsafe {
            ffi::scn_hit_test_result_world_coordinates(self.ptr, value.as_mut_ptr().cast())
        };
        ok.then_some(value)
    }
}

impl View {
    /// Mirrors `SCNView.hitTest`.
    #[must_use]
    pub fn hit_test(&self, point: crate::CGPoint) -> Option<HitTestResults> {
        unsafe { HitTestResults::from_raw(ffi::scn_view_hit_test(self.ptr, point.x, point.y)) }
    }
}
